File: An2/MAP/MAP-C#/MAP-Lab6/MAP-Lab6/Repository/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using MAP_Lab6.Utils.LinkedList;
using MAP_Lab6.Utils;

namespace MAP_Lab6.Repository
{
    class Stack<T>
    {
        private MAP_Lab6.Utils.LinkedList.LinkedList<T> elements;
        private int size = 0;

        public Stack()
        {
            elements = new MAP_Lab6.Utils.LinkedList.LinkedList<T>();
        }

        public void push(T data)
        {
            Node<T> lastNode = this.elements.getLastElement();
            Node<T> nodeToAdd = new Node<T>(data);

            if (lastNode != null)
            {
                lastNode.next = nodeToAdd;
            }
            else
            {
                this.elements.firstNode = nodeToAdd;
            }
            size++;
        }

        public T pop()
        {
            if (this.size == 0)
            {
                throw new MyException("Stack is empty!");
            }
            this.size--;
            Node<T> lastNode = this.elements.getLastElement();
            T returnData = lastNode.data;
            this.elements.removeNode(lastNode);
            return returnData;
        }

        public int getSize()
        {
            return this.size;
        }

        public Boolean isEmpty()
        {
            return (this.size == 0);
        }

        public Stack<T> copy()
        {
            Stack<T> copy = new Stack<T>();
            copy.elements = this.elements.copy();
            copy.size = this.size;

            return copy;
        }
    }
}

File: An2/Sem1/MAP/MAP-Java/MAP-Lab6/src/Repository/Stack.java
package Repository;

import Utils.LinkedList.LinkedList;
import Utils.LinkedList.Node;
import Utils.MyException;

/**
 * Created with IntelliJ IDEA.
 * User: mihai
 * Date: 11/27/13
 * Time: 2:24 PM
 * To change this template use File | Settings | File Templates.
 */
public class Stack<T> {
    private int size = 0;
    private LinkedList<T> elements;

    public Stack(){
        elements = new LinkedList<T>();
    }

    public void push(T data){
        Node<T> lastNode = elements.getLastElement();
        Node<T> nodeToAdd = new Node<T>(data);

        if (lastNode != null)
            lastNode.next = nodeToAdd;
        else
            elements.firstNode = nodeToAdd;
        size++;
    }

    public T pop() throws MyException {
        if (this.isEmpty())
            throw new MyException("Cannot pop element from empty stack!");
        else {
            size--;
            Node<T> lastNode = elements.getLastElement();
            T returnData = lastNode.data;
            elements.removeNode(lastNode);
            return returnData;
        }
    }

    public int getStackSize(){
        return this.size;
    }

    public boolean isEmpty(){
        return this.size < 1;
    }

    public Stack<T> copy() {
        Stack<T> copy = new Stack();
        copy.elements = this.elements.copy();
        copy.size = this.size;

        return copy;
    }

}
